From afabfff06e2dece1772ed788ac41ca0d297ab49b Mon Sep 17 00:00:00 2001
From: Lukas Schauer <lukas@schauer.so>
Date: Mon, 1 Aug 2016 20:35:46 +0200
Subject: [PATCH] updated url to letsencrypt license agreement

--- a/letsencrypt.sh
+++ b/letsencrypt.sh
@@ -68,7 +68,7 @@
 
   # Default values
   CA="https://acme-v01.api.letsencrypt.org/directory"
-  LICENSE="https://letsencrypt.org/documents/LE-SA-v1.0.1-July-27-2015.pdf"
+  LICENSE="https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf"
   CHALLENGETYPE="http-01"
   CONFIG_D=
   HOOK=
--- a/docs/examples/config.sh.example
+++ b/docs/examples/config.sh.example
@@ -15,8 +15,8 @@
 # Path to certificate authority (default: https://acme-v01.api.letsencrypt.org/directory)
 #CA="https://acme-v01.api.letsencrypt.org/directory"
 
-# Path to license agreement (default: https://letsencrypt.org/documents/LE-SA-v1.0.1-July-27-2015.pdf)
-#LICENSE="https://letsencrypt.org/documents/LE-SA-v1.0.1-July-27-2015.pdf"
+# Path to license agreement (default: https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf)
+#LICENSE="https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf"
 
 # Which challenge should be used? Currently http-01 and dns-01 are supported
 #CHALLENGETYPE="http-01"
